File: src/agentsfleetd/fleet/budget.zig
//! Per-fleet spend ceilings — the `daily_dollars` / `monthly_dollars` a fleet
//! author declares in `TRIGGER.md`, enforced.
//!
//! Two gates consume this module, mirroring the two places the platform already
//! asks "may this run proceed?":
//!   - `service_billing.runBilling` — before the receive debit, so a refused
//!     event is never charged. Refusal writes `gate_blocked` + `budget_breach`.
//!   - `service_renew.renew` — the mid-run kill. Refusal answers `UZ-RUN-015`,
//!     the runner terminates its child, and the report lands `fleet_error` +
//!     `budget_breach`.
//!
//! The ceiling is a floor-check, not a projection: a run is admitted while
//! `spend < cap`. An admitted run may overshoot by at most one renewal window's
//! worth of tokens before its next `/renew` refuses it — bounded, not unbounded.
//!
//! Spend means credit DRAINED (`credit_deducted_nanos`), not credit metered: on
//! the slice that exhausts a wallet the remainder is forgiven, and a budget must
//! count money that actually left the pool.
//!
//! Both windows derive from ONE caller-supplied `now_ms` (RULE TIM) — this
//! module never reads the clock, so the day and month windows can never straddle
//! a tick, and every test pins time by argument.
//!
//! Fail-open posture: a DB fault yields `null`, and both callers admit the run.
//! This mirrors `metering.balanceCoversEstimate` ("Any DB failure returns true
//! (fail-open) so the gate never turns into an availability incident"). A budget
//! gate stricter than the platform's own credit gate would be an inconsistent
//! guarantee. An *unparseable stored budget* is a different thing entirely and
//! fails CLOSED — a ceiling we cannot read is not a ceiling we may ignore.

const std = @import("std");
const sql = @import("sql.zig");
const pg = @import("pg");
const logging = @import("log");
const common = @import("common");
const clock = common.clock;

const PgQuery = @import("../db/pg_query.zig").PgQuery;
const ec = @import("../errors/error_registry.zig");
const tenant_billing = @import("../state/tenant_billing.zig");
const ChargeType = @import("../state/fleet_telemetry_store.zig").ChargeType;
const config_types = @import("../fleet_runtime/config_types.zig");
const config_helpers = @import("../fleet_runtime/config_helpers.zig");

const log = logging.scoped(.fleet_budget);

pub const FleetBudget = config_types.FleetBudget;

/// The rolling daily window. Not a calendar day: `authoring.mdx` documents
/// "Rolling 24-hour dollar ceiling", so the window slides with `now_ms`.
const ROLLING_DAY_MS: i64 = std.time.ms_per_day;

/// Why these queries sum `metering_periods`, not the telemetry stage row: a run
/// drains slice-by-slice across up to `MAX_RUNTIME_MS`, but the accumulating
/// telemetry STAGE row pins `recorded_at` at the first renewal (~run start) and
/// never advances it, so it cannot answer "how much drained in the last 24h".
/// The accurate per-slice times live in `metering_periods.created_at`; the stage
/// telemetry row is joined only for fleet scope + index pruning. The receive fee
/// is one telemetry row written once at gate-pass, so ITS `recorded_at` is
/// already the drain time. Pruning: a slice draining in `[floor, now]` cannot
/// belong to a run started before `floor - MAX_RUNTIME`, so bounding the stage
/// scan there drops no slice while keeping the index range small.
const DRAIN_BACKOFF_MS: i64 = common.MAX_RUNTIME_MS;

/// Credit drained by one fleet inside each window, in nanos.
pub const Spend = struct {
    day_nanos: i64,
    month_nanos: i64,
};

/// Why the gate refused — the caller maps this onto a log line, never onto a
/// distinct wire code (both ceilings are one `budget_breach` to the operator).
pub const Verdict = enum {
    ok,
    day_exceeded,
    month_exceeded,

    pub fn refused(self: Verdict) bool {
        return self != .ok;
    }
};

/// A budget the renew gate could not read. Distinct from "could not reach the
/// database": a malformed stored budget fails CLOSED.
pub const BudgetError = error{UnreadableBudget};

/// A fleet's stored ceiling paired with what it has already spent.
pub const BudgetAndSpend = struct { budget: FleetBudget, spend: Spend };

/// What the renew-side read observed. Splitting "we read nothing" into three
/// distinct causes is what lets the two failure postures be *decided* by a pure
/// function (`refusalFor`) rather than buried in a `catch` beside a connection.
pub const BudgetRead = union(enum) {
    found: BudgetAndSpend,
    /// Nothing to enforce: no fleet row (the lease's own checks own that case),
    /// or a fleet that declares no `budget` at all. Undeclared is unbounded —
    /// killing such a run would enforce a ceiling nobody wrote.
    absent,
    /// A budget IS declared but cannot be parsed — fail CLOSED.
    unreadable,
    /// The database could not be reached or queried — fail OPEN.
    unavailable,
};

/// The mid-run refusal decision. Returns the breach verdict when the run must
/// stop, `null` when it may continue.
///
/// The whole asymmetry lives here, in one switch, testable without a database:
/// an *unavailable* budget admits (a metering outage must not kill every
/// in-flight run), an *unreadable* one refuses (a ceiling we cannot parse is not
/// a ceiling we may ignore).
pub fn refusalFor(read: BudgetRead) ?Verdict {
    return switch (read) {
        .found => |f| {
            const verdict = covers(f.budget, f.spend);
            return if (verdict.refused()) verdict else null;
        },
        .absent, .unavailable => null,
        .unreadable => .day_exceeded,
    };
}

/// The pre-run admission decision. A spend we could not read admits the event
/// (fail open), mirroring `metering.balanceCoversEstimate`.
pub fn verdictOrAdmit(maybe_spend: ?Spend, fleet_budget: FleetBudget) Verdict {
    const spend = maybe_spend orelse return .ok;
    return covers(fleet_budget, spend);
}

/// Dollars → nanos. The parser bounds inputs to `(0, 1000]` daily and
/// `(0, 10000]` monthly, so `dollars * 1e9` tops out around 1e13 and cannot
/// approach `i64` max; the saturating branch guards a future bound change, not
/// a live path. Rounds to nearest so `0.000000001` is one nano rather than zero
/// (a truncating cast would silently make tiny budgets free).
///
/// A non-finite or non-positive ceiling collapses to ZERO, not to "unlimited":
/// `covers` then refuses every run, because a budget we cannot make sense of
/// must not silently become permission to spend without bound.
pub fn dollarsToNanos(dollars: f64) i64 {
    if (!std.math.isFinite(dollars) or dollars <= 0) return 0;
    const scaled = @round(dollars * @as(f64, @floatFromInt(tenant_billing.NANOS_PER_USD)));
    if (scaled >= @as(f64, @floatFromInt(std.math.maxInt(i64)))) return std.math.maxInt(i64);
    return @intFromFloat(scaled);
}

/// The ceiling comparison. Admitted while strictly under; refused at equality,
/// so a fleet that has spent exactly its `daily_dollars` runs no further.
/// `monthly_dollars` is optional — absent means no monthly ceiling.
pub fn covers(budget: FleetBudget, spend: Spend) Verdict {
    if (spend.day_nanos >= dollarsToNanos(budget.daily_dollars)) return .day_exceeded;
    if (budget.monthly_dollars) |monthly| {
        if (spend.month_nanos >= dollarsToNanos(monthly)) return .month_exceeded;
    }
    return .ok;
}

/// Window floors for one `now_ms`. Split out so both queries and every test
/// derive their bounds from the same arithmetic.
pub fn windowFloors(now_ms: i64) struct { day: i64, month: i64 } {
    return .{
        .day = now_ms -| ROLLING_DAY_MS,
        .month = clock.startOfUtcMonthMillis(now_ms),
    };
}

/// Credit drained by `fleet_id` inside each window. `null` on ANY database
/// fault — the caller then fails open explicitly rather than by coincidence.
pub fn spendForFleet(
    pool: *pg.Pool,
    workspace_id: []const u8,
    fleet_id: []const u8,
    now_ms: i64,
) ?Spend {
    const conn = pool.acquire() catch |err| {
        log.warn("budget_acquire_fail", .{ .error_code = ec.ERR_INTERNAL_OPERATION_FAILED, .fleet_id = fleet_id, .err = @errorName(err) });
        return null;
    };
    defer pool.release(conn);
    return spendForFleetOn(conn, workspace_id, fleet_id, now_ms) catch |err| {
        log.warn("budget_spend_query_failed", .{ .error_code = ec.ERR_INTERNAL_DB_QUERY, .fleet_id = fleet_id, .err = @errorName(err) });
        return null;
    };
}

/// The connection-taking half of `spendForFleet`, so a caller that already holds
/// a connection (the renew gate, the integration tests) does not take a second
/// one from the pool. A fleet with no telemetry rows yet spends zero.
pub fn spendForFleetOn(conn: *pg.Conn, workspace_id: []const u8, fleet_id: []const u8, now_ms: i64) !?Spend {
    const floors = windowFloors(now_ms);
    var q = PgQuery.from(try conn.query(sql.SELECT_BUDGET_DRAIN, .{
        workspace_id,             fleet_id,                   floors.day,       floors.month,
        ChargeType.stage.label(), ChargeType.receive.label(), DRAIN_BACKOFF_MS,
    }));
    defer q.deinit();
    const row = try q.next() orelse return Spend{ .day_nanos = 0, .month_nanos = 0 };
    return Spend{
        .day_nanos = try row.get(i64, 0),
        .month_nanos = try row.get(i64, 1),
    };
}

/// Classify the renew-side read into the three no-budget causes plus the happy
/// path, so `refusalFor` can decide without touching a connection.
pub fn readBudget(
    conn: *pg.Conn,
    alloc: std.mem.Allocator,
    fleet_id: []const u8,
    workspace_id: []const u8,
    now_ms: i64,
) BudgetRead {
    const found = fetchBudgetAndSpend(conn, alloc, fleet_id, workspace_id, now_ms) catch |err| {
        if (err == BudgetError.UnreadableBudget) return .unreadable;
        log.warn("budget_read_query_failed", .{ .error_code = ec.ERR_INTERNAL_DB_QUERY, .fleet_id = fleet_id, .err = @errorName(err) });
        return .unavailable;
    } orelse return .absent;
    return .{ .found = found };
}

/// Renew-side read. Returns `null` when there is no ceiling to enforce — either
/// the fleet row is gone (the lease's own checks own that case) or the fleet
/// declares no `budget` at all. Returns `BudgetError.UnreadableBudget` when a
/// budget IS declared but cannot be parsed (fail CLOSED), and propagates DB
/// errors so the caller can fail open on them.
///
/// The absent/malformed split is load-bearing. `parseFleetConfig` requires
/// `budget`, so a live fleet always has one — but `config_json` rows written by
/// other paths (fixtures, anything predating the requirement) may not, and
/// refusing THEIR renewals would kill healthy in-flight runs to enforce a
/// ceiling nobody declared. Undeclared means unbounded here, exactly as it was
/// before this gate existed; the tenant credit pool still bounds it.
pub fn fetchBudgetAndSpend(
    conn: *pg.Conn,
    alloc: std.mem.Allocator,
    fleet_id: []const u8,
    workspace_id: []const u8,
    now_ms: i64,
) !?BudgetAndSpend {
    const floors = windowFloors(now_ms);
    var q = PgQuery.from(try conn.query(sql.SELECT_BUDGET_POLICY_AND_DRAIN, .{
        fleet_id,                 workspace_id,               fleet_id,         floors.day, floors.month,
        ChargeType.stage.label(), ChargeType.receive.label(), DRAIN_BACKOFF_MS,
    }));
    defer q.deinit();
    const row = try q.next() orelse return null; // no fleet row
    // SQL NULL here means the JSON path found no `budget` key: no ceiling was
    // declared, so there is nothing to enforce. A DECLARED-but-malformed budget
    // is a different thing and fails closed inside `parseStoredBudget`.
    const budget_json = try row.get(?[]const u8, 0) orelse return null;
    const spend = Spend{
        .day_nanos = try row.get(i64, 1),
        .month_nanos = try row.get(i64, 2),
    };
    // `null` here = "not a declared ceiling" (JSON null, or a non-object in the
    // budget slot) → admit, exactly like a missing key. `error` = "a ceiling was
    // declared as an object and it will not validate" → fail closed.
    const parsed_budget = try parseStoredBudget(alloc, budget_json) orelse return null;
    return .{ .budget = parsed_budget, .spend = spend };
}

/// Parse the stored budget subobject through the SAME validator that accepted it
/// at ingest (`config_helpers.parseFleetBudget`), so the ceiling that admits a
/// run and the ceiling that kills it can never be interpreted two ways.
///
/// Three outcomes: `null` = not a budget OBJECT (JSON null / scalar / array) →
/// no ceiling declared → admit, like a missing key (a present-but-null key
/// renders as the text `"null"`, so without this it would fail closed and kill
/// in-flight runs). `error.UnreadableBudget` = a budget OBJECT that won't
/// validate → someone botched a ceiling → fail closed. A `FleetBudget` = valid.
pub fn parseStoredBudget(alloc: std.mem.Allocator, budget_json: []const u8) !?FleetBudget {
    const parsed = std.json.parseFromSlice(std.json.Value, alloc, budget_json, .{}) catch
        return BudgetError.UnreadableBudget;
    defer parsed.deinit();
    const obj = switch (parsed.value) {
        .object => |o| o,
        else => return null, // not a declared ceiling → admit
    };
    return config_helpers.parseFleetBudget(obj) catch BudgetError.UnreadableBudget;
}

File: rustd/crates/afd_fleet/src/money/budget.rs
//! The fleet's own spend ceiling, and what it has already drained against it.
//!
//! # Spend means money that LEFT the pool
//!
//! Not money metered. On the slice that exhausts a wallet the remainder is
//! forgiven, and a ceiling must count what a tenant was actually charged —
//! otherwise a fleet that ran out of credit would keep accruing against a cap
//! it never spent.
//!
//! # The ceiling is a floor-check, not a projection
//!
//! A run is admitted while `spend < cap`, and refused AT equality. An admitted
//! run may overshoot by at most one renewal window before its next `/renew`
//! refuses it — bounded, not unbounded, and that bound is the design rather
//! than an accident.
//!
//! # Two failures that look alike and point opposite ways
//!
//! `budget.zig` is the module that got this right, and this is its rule kept
//! rather than restated: a ceiling we could not READ is not a ceiling we may
//! ignore, but a datastore we could not REACH must not kill every fleet on the
//! platform. So an unreadable budget fails CLOSED and an unavailable one fails
//! OPEN.
//!
//! Here that asymmetry needs no union type, because Rust already has the two
//! channels: a fault is `Err`, and a verdict — including "the stored ceiling
//! will not parse" — is `Ok`. The Zig needs `BudgetRead` with four arms
//! precisely because it has only one channel to say both things through.

use afd_core::clock::UnixMillis;
use afd_core::id::Uuid7;
use afd_fleet_runtime::config::Budget;
use sqlx::Row as _;

use crate::error::{Result, query};
use crate::money::Nanos;
use crate::money::store::Accounts;
use crate::money::window::Windows;
use crate::sql;

/// Statement name, for the context a query failure carries.
const CONTEXT_DRAIN: &str = "fleet budget drain";

/// What a fleet has drained inside each window.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct Spend {
    /// Drained inside the rolling twenty-four hours.
    pub day: Nanos,
    /// Drained since the UTC month opened.
    pub month: Nanos,
}

/// Whether a fleet may start more work, and which ceiling stopped it.
///
/// The caller maps this onto a log line, never onto a distinct wire code: both
/// ceilings are one `budget_breach` to the operator, and splitting them on the
/// wire would make a dashboard filter on a distinction nobody acts on.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Verdict {
    /// Under both ceilings.
    Admit,
    /// The rolling-day ceiling is reached.
    DayExceeded,
    /// The calendar-month ceiling is reached.
    MonthExceeded,
}

impl Verdict {
    /// Whether this verdict stops the work.
    #[must_use]
    pub const fn refuses(self) -> bool {
        !matches!(self, Self::Admit)
    }

    /// The spelling a log line carries.
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Admit => "admit",
            Self::DayExceeded => "day_exceeded",
            Self::MonthExceeded => "month_exceeded",
        }
    }
}

/// Whether `budget` covers `spend`.
///
/// Pure and total — no clock, no connection, no allocation — which is what
/// lets every ceiling case be proven in a unit test rather than against a
/// seeded database. `budget.zig` makes the same point about its own `covers`
/// and then reaches for a `*pg.Conn` two functions later; here the boundary is
/// the function signature.
///
/// The day ceiling is checked first because it is the one that always exists:
/// `monthly_dollars` is optional, and an absent monthly ceiling means no
/// monthly ceiling — not an implicit one derived from the daily.
#[must_use]
pub fn covers(budget: Budget, spend: Spend) -> Verdict {
    if spend.day.has_reached(Nanos::from_dollars(budget.daily())) {
        return Verdict::DayExceeded;
    }
    match budget.monthly() {
        Some(monthly) if spend.month.has_reached(Nanos::from_dollars(monthly)) => {
            Verdict::MonthExceeded
        }
        _ => Verdict::Admit,
    }
}

impl Accounts {
    /// What `fleet_id` has drained inside the windows opening at `now`.
    ///
    /// A fleet with no ledger rows spends zero — the statement's `COALESCE`
    /// answers that, so there is no empty-result branch here.
    ///
    /// # Errors
    /// Reports a datastore that would not answer. The caller applies the
    /// fail-open posture; this does not decide it, which is the whole point of
    /// returning `Result` rather than an `Option` that has already chosen.
    pub async fn spend(
        &self,
        workspace_id: &Uuid7,
        fleet_id: &Uuid7,
        now: UnixMillis,
    ) -> Result<Spend> {
        let windows = Windows::at(now);
        let mut connection = self.pool().acquire().await?;
        let row = sqlx::query(sql::billing::SELECT_BUDGET_DRAIN)
            .bind(workspace_id.as_str())
            .bind(fleet_id.as_str())
            .bind(windows.day.as_millis())
            .bind(windows.month.as_millis())
            .bind(sql::billing::charge::STAGE)
            .bind(sql::billing::charge::RECEIVE)
            .fetch_one(&mut *connection)
            .await
            .map_err(query(CONTEXT_DRAIN))?;

        Ok(Spend {
            day: Nanos::from_i64(row.try_get(0).map_err(query(CONTEXT_DRAIN))?),
            month: Nanos::from_i64(row.try_get(1).map_err(query(CONTEXT_DRAIN))?),
        })
    }
}

#[cfg(test)]
mod tests {
    #![expect(
        clippy::expect_used,
        reason = "a test asserts by panicking; the manifest's restriction set is for the daemon"
    )]
    use super::{Spend, Verdict, covers};
    use crate::money::{NANOS_PER_USD, Nanos};
    use afd_fleet_runtime::config::Budget;

    /// A ceiling built the only way one can be: through the parser that bounds
    /// it.
    ///
    /// There is no `Budget` constructor, deliberately — a ceiling that did not
    /// come through validation could be negative, infinite, or above the cap,
    /// and every rule in this file assumes it is none of those. So even a unit
    /// test goes the long way round, which is the type doing its job.
    fn ceiling(daily: f64, monthly: Option<f64>) -> Budget {
        let monthly = monthly.map_or(String::new(), |m| format!(r#","monthly_dollars":{m}"#));
        let document = format!(
            r#"{{"name":"b","x-agentsfleet":{{"triggers":[{{"type":"api"}}],"tools":[],"budget":{{"daily_dollars":{daily}{monthly}}}}}}}"#
        );
        afd_fleet_runtime::FleetConfig::stored(&document)
            .expect("the fixture document is well-formed")
            .budget()
    }

    #[test]
    fn a_fleet_under_both_ceilings_is_admitted() {
        // One dollar today against a five-dollar cap, two this month against
        // ten — written in the unit the ceiling is authored in rather than as
        // raw nanos, so the comparison being tested is legible without
        // counting zeroes.
        let spend = Spend {
            day: Nanos::from_i64(NANOS_PER_USD),
            month: Nanos::from_i64(2 * NANOS_PER_USD),
        };
        assert_eq!(covers(ceiling(5.0, Some(10.0)), spend), Verdict::Admit);
    }

    #[test]
    fn the_day_ceiling_refuses_at_equality_not_past_it() {
        // Exactly five dollars drained against a five-dollar day. The Zig
        // spells this `spend >= cap`, so a fleet that has spent precisely its
        // ceiling runs no further — one more run would be the first dollar of
        // an overdraft nobody authorised.
        let spend = Spend {
            day: Nanos::from_i64(5_000_000_000),
            month: Nanos::ZERO,
        };
        assert_eq!(covers(ceiling(5.0, None), spend), Verdict::DayExceeded);
    }

    #[test]
    fn an_absent_monthly_ceiling_is_no_ceiling_rather_than_a_derived_one() {
        // A year of spend against a fleet that declared only a daily cap. The
        // month must not be inferred from the day — an implicit 30× ceiling
        // would refuse runs the author never limited.
        let spend = Spend {
            day: Nanos::ZERO,
            month: Nanos::from_i64(900_000_000_000),
        };
        assert_eq!(covers(ceiling(5.0, None), spend), Verdict::Admit);
    }

    #[test]
    fn the_month_ceiling_refuses_when_the_day_still_has_room() {
        // The case an order-dependent check would miss: today is quiet, the
        // month is spent.
        let spend = Spend {
            day: Nanos::ZERO,
            month: Nanos::from_i64(10_000_000_000),
        };
        assert_eq!(
            covers(ceiling(5.0, Some(10.0)), spend),
            Verdict::MonthExceeded
        );
    }

    #[test]
    fn a_refusal_is_a_refusal_whichever_ceiling_produced_it() {
        assert!(!Verdict::Admit.refuses());
        assert!(Verdict::DayExceeded.refuses());
        assert!(Verdict::MonthExceeded.refuses());
    }
}
